File: leetcode/q219.go
package leetcode

// 219. Contains Duplicate II

// TIME: O(n)
// LeetCode: Runtime: 146 ms - Memory: 17.7 MB
func containsNearbyDuplicate(nums []int, k int) bool {
	occurrence := make(map[int]int, 0)
	exist := make(map[int]bool, 0)
	for i, num := range nums {
		if exist[num] && (i-occurrence[num] <= k) {
			return true
		}
		occurrence[num] = i
		exist[num] = true
	}
	return false
}

File: src/q219.rs
use std::collections::HashMap;

pub struct Solution {}

impl Solution {
    // LEETCODE: RunTime: 52ms - Memory: 6.9MB
    pub fn contains_nearby_duplicate(nums: Vec<i32>, k: i32) -> bool {
        if k == 0 {
            return false;
        }
        let mut occur_map = HashMap::new();
        let ref_nums = &nums;
        for (i, num) in ref_nums.into_iter().enumerate() {
            if occur_map.contains_key(num) && (i - occur_map[num] <= k as usize) {
                return true;
            }
            occur_map.insert(num, i);
        }
        return false;
    }
}
